compiler/rustc_arena/src/lib.rs

```diff
@@ -133,6 +160,10 @@ where
     }
 }
 
+#[rustc_unsafe_specialization_marker]
+trait Clonable: Clone {}
+impl<T: Clone> Clonable for T {}
```

Comment:
My somewhat limited understanding of the `min_specialization` rules is that you're here promising that any specialization relying on `Clonable` will only be applied to types where `T: Clone` is not dependent lifetimes (directly or indirectly)
E.g. that there's no `impl Clone for &'static Foo` involved. That can only be guaranteed if you could also manually enumerate all T.

It would also mean that it's unsafe to use `TypedArena` for such types and users of it would have to uphold that in the future.

Reply:
I'm not quite sure if this is the case, but I don't know the feature well enough. I was going by the example of `FusedIterator` and `TrustedLen`, both of which are implemented for e.g. `std::slice::Iterator<'a, T: 'a>` and I believe both are specialization marker traits.

Reply:
The thing is that `FusedIterator` or `TrustedLen` are not implemented for arbitrary `T`. They're only implemented for structs controlled by std and those `impl` are only conditional on additional bounds that themselves are other specialization traits.

What's different here is that you're essentially declaring that any and all `Clone` types can act as a specialization marker for specializations relying on `Clonable`. This would only be valid when you have some way of reasoning about the set of `T`. And since `T` is what goes into `TypedArena` it means you've ensured that all uses of `TypedArena` are compatible with the min specialization rules. This kind of far-reaching obligation is why there's the word `unsafe` in that annotation.

But again, my understanding of `min_specialization` is quite limited, so I may be wrong here.

Reply:
All right, makes sense. I wish I had another way to specialize for `Cloned<slice::Iter>` :(